src/main/scala/org/broadinstitute/dsde/firecloud/dataaccess/ElasticSearchTrialDAO.scala

```diff
@@ -0,0 +1,208 @@
+    val insert = client
+      .prepareIndex(indexName, datatype, projectName.value)
+      .setSource(trialProject.toJson.compactPrint, XContentType.JSON)
+      .setCreate(true) // fail the request if the project already exists
+      .setRefreshPolicy(refreshMode)
+
+    executeESRequest[IndexRequest, IndexResponse, IndexRequestBuilder](insert) // will throw error if insert fails
+    trialProject
+  }
+
+  /**
+    * Update the "verified" field for a specified project record. The "verified" field indicates whether
+    * or not the associated billing project was created successfully in Google Cloud. Throws an error if
+    * the record was not found or the record could not be updated.
+    *
+    * @param projectName name of the project record to update
+    * @param verified verified value with which to update the project record
+    * @return the updated project record
+    */
+  override def verifyProject(projectName: RawlsBillingProjectName, verified: Boolean): TrialProject = {
+    val (version, project) = getProjectInternal(projectName)
+
+    if (project.verified == verified) {
+      project
+    } else {
+      val updatedProject = project.copy(verified = verified)
+      val update = client
+        .prepareUpdate(indexName, datatype, updatedProject.name.value)
+        .setDoc(updatedProject.toJson.compactPrint, XContentType.JSON)
+        .setVersion(version) // guarantee nobody else has updated in the meantime
+        .setRefreshPolicy(refreshMode)
+
+      executeESRequest[UpdateRequest, UpdateResponse, UpdateRequestBuilder](update) // will throw error if update fails
+      updatedProject
+    }
+  }
+
+  /**
+    * Associates the next-available project record with a specified user. "Next available"
+    * is defined as verified, unclaimed, and first alphabetically by project name.
+    * Throws an error if no project records are available, or if the project
+    * record could not be updated.
+    *
+    * @param userInfo the user (email and subjectid) with which to update the project record.
+    * @return the updated project record
+    */
+  override def claimProject(userInfo: WorkbenchUserInfo): TrialProject = {
+    val nextProjectQuery = boolQuery()
+      .must(termQuery("verified", true))
+      .mustNot(existsQuery("user.userSubjectId.keyword"))
+
+    // if we find regular race conditions in which multiple users attempt to claim the "next" project,
+    // we could change this to return N (= ~20) available projects, then choose a random project
+    // from that list.
+    val nextProjectRequest = client
+      .prepareSearch(indexName)
+      .setQuery(nextProjectQuery)
+      .addSort("name.keyword", SortOrder.ASC)
+      .setSize(1)
+      .setVersion(true)
+
+    val nextProjectResponse = executeESRequest[SearchRequest, SearchResponse, SearchRequestBuilder](nextProjectRequest)
+
+    if (nextProjectResponse.getHits.totalHits == 0)
+      throw new FireCloudException("no available projects")
```

Comment:
I'm guessing it's left to the calling code (that doesn't exist yet) to catch this exception, create a new BP, and immediately claim it? Would it make sense to have a companion method `claimProject(project: TrialProject, userInfo: WorkbenchUserInfo)`?

Maybe we distrust BP creation enough that it has to be done in a `while` loop, and a "second chance" codepath isn't good enough.

Reply:
correct, I was leaving it up to calling code to catch & handle any exceptions. Since this is prospective development and we don't know exactly what the calling code will need, I didn't build out too much ... I could certainly see changing/adding methods to the dao once we really understand the use cases.